cgen: Free optional string

### Describe the bug

Try to free an structure with a optional string

### Reproduction Steps

```v
module main

pub struct Test {
    b ?string
}

fn main() {
    t := Test{
        b: "b"
    }

    println("t: ${t}")

    defer {
        unsafe {
            t.free()
        }
    }
}
```

### Expected Behavior

The program compile and run correctly without errors

### Current Behavior

```bash
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/jsonx.01J5G50V9VFVJAHVS165MZ904H.tmp.c:2223: warning: implicit declaration of function '_option_string_free'
cc: /home/esquerbatua/git/v/thirdparty/tcc/lib/libgc.a: error: 'GC_noop1_ptr' defined twice
cc: tcc: error: undefined symbol '_option_string_free'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

Current V version: V 0.4.7 e9c9580, timestamp: 2024-08-17 14:34:42 +0300

### Environment details (OS name and version, etc.)

V full version: V 0.4.7 4940f2d.e9c9580
OS: linux, Ubuntu 24.04 LTS
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 7840HS w/ Radeon 780M Graphics

vexe: /home/esquerbatua/git/v/v
vexe mtime: 2024-08-17 12:30:16

vroot: OK, value: /home/esquerbatua/git/v
VMODULES: OK, value: /home/esquerbatua/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.43.0
Git vroot status: weekly.2024.33-19-ge9c95803
.git/config present: true

CC version: cc (Ubuntu 13.2.0-23ubuntu4) 13.2.0
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/auto_free_methods.v
+++ b/vlib/v/gen/c/auto_free_methods.v
@@ -30,7 +30,7 @@ fn (mut g Gen) gen_free_methods() {
 fn (mut g Gen) gen_free_method(typ ast.Type) string {
 	styp := g.typ(typ).replace('*', '')
 	mut fn_name := styp_to_free_fn_name(styp)
-	deref_typ := typ.set_nr_muls(0)
+	deref_typ := if typ.has_flag(.option) { typ } else { typ.set_nr_muls(0) }
 	if deref_typ in g.generated_free_methods {
 		return fn_name
 	}
@@ -81,7 +81,8 @@ fn (mut g Gen) gen_free_for_struct(typ ast.Type, info ast.Struct, styp string, f
 		if sym.kind !in [.string, .array, .map, .struct_] {
 			continue
 		}
-		mut field_styp := g.typ(field.typ).replace('*', '')
+		mut field_styp := g.typ(field.typ.set_nr_muls(0).clear_flag(.option)).replace('*',
+			'')
 		is_shared := field_styp.starts_with('__shared')
 		is_struct_option := typ.has_flag(.option)
 		if is_shared {
@@ -92,14 +93,34 @@ fn (mut g Gen) gen_free_for_struct(typ ast.Type, info ast.Struct, styp string, f
 		} else {
 			g.gen_free_method(field.typ)
 		}
+		is_field_option := field.typ.has_flag(.option)
+		expects_opt := field_styp_fn_name.starts_with('_option_')
 		if is_shared {
 			fn_builder.writeln('\t${field_styp_fn_name}(&(it->${field_name}->val));')
 		} else if is_struct_option {
 			opt_styp := g.base_type(typ)
 			prefix := if field.typ.is_ptr() { '' } else { '&' }
-			fn_builder.writeln('\t${field_styp_fn_name}(${prefix}((${opt_styp}*)&it->data)->${field_name});')
+			if is_field_option {
+				opt_field_styp := if expects_opt { g.typ(field.typ) } else { g.base_type(field.typ) }
+				suffix := if expects_opt { '' } else { '.data' }
+
+				fn_builder.writeln('\tif (((${opt_styp}*)&it->data)->${field_name}.state != 2) {')
+				fn_builder.writeln('\t\t${field_styp_fn_name}((${opt_field_styp}*)${prefix}((${opt_styp}*)&it->data)->${field_name}${suffix});')
+				fn_builder.writeln('\t}')
+			} else {
+				fn_builder.writeln('\t${field_styp_fn_name}(${prefix}((${opt_styp}*)&it->data)->${field_name});')
+			}
 		} else {
-			fn_builder.writeln('\t${field_styp_fn_name}(&(it->${field_name}));')
+			if is_field_option {
+				opt_field_styp := if expects_opt { g.typ(field.typ) } else { g.base_type(field.typ) }
+				suffix := if expects_opt { '' } else { '.data' }
+
+				fn_builder.writeln('\tif (it->${field_name}.state != 2) {')
+				fn_builder.writeln('\t\t${field_styp_fn_name}((${opt_field_styp}*)&(it->${field_name}${suffix}));')
+				fn_builder.writeln('\t}')
+			} else {
+				fn_builder.writeln('\t${field_styp_fn_name}(&(it->${field_name}));')
+			}
 		}
 	}
 	fn_builder.writeln('}')

--- a/vlib/v/tests/option_free_method_test.v
+++ b/vlib/v/tests/option_free_method_test.v
@@ -9,6 +9,7 @@ pub struct Test {
 	d ?&Test2
 }
 
+@[manualfree]
 fn test_main() {
 	t := Test{
 		b: 'b'

--- a/vlib/v/tests/option_free_method_test.v
+++ b/vlib/v/tests/option_free_method_test.v
@@ -1,0 +1,35 @@
+pub struct Test2 {
+	a int
+}
+
+pub struct Test {
+	a Test2
+	b ?string
+	c ?Test2
+	d ?&Test2
+}
+
+fn test_main() {
+	t := Test{
+		b: 'b'
+	}
+
+	println('t: ${t}')
+
+	defer {
+		unsafe {
+			t.free()
+		}
+	}
+
+	mut t2 := ?Test(Test{
+		b: 'b'
+	})
+	println('t: ${t2}')
+
+	defer {
+		unsafe {
+			t2?.free()
+		}
+	}
+}

